### PR TITLE
Issue #12443: Fix null pointer exception on config value with trailin…

### LINKED ISSUE #12443: checkstyle 10.3.3 and newer throws NullPointerException on annotated generic types from IllegalType rule
I minimized my repro to narrow down the cause.

The config:

`❯ /bin/cat checkstyle-configuration.xml`

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="IllegalType">
            <property name="illegalClassNames"
                    value="java.lang.StringBuffer, " />
        </module>
    </module>
</module>
```

The code:

`❯ cat ExampleClass.java`

```java
package com.example;

import java.util.List;

import javax.validation.constraints.Pattern;

public class ExampleClass
{
    public void example(List<@Pattern String> strings)
    {
    }
}
```

```bash
❯ RUN_LOCALE="-Duser.language=en -Duser.country=US"
❯ java $RUN_LOCALE -jar checkstyle-10.3.3-all.jar -c checkstyle-configuration.xml ExampleClass.java 
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing ExampleClass.java
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:307)
	at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:224)
	at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:416)
	at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:339)
	at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:196)
	at com.puppycrawl.tools.checkstyle.Main.main(Main.java:131)
Caused by: java.lang.NullPointerException: Cannot invoke "com.puppycrawl.tools.checkstyle.api.DetailAST.getLineNo()" because "ast" is null
	at com.puppycrawl.tools.checkstyle.api.AbstractCheck.log(AbstractCheck.java:276)
	at com.puppycrawl.tools.checkstyle.checks.coding.IllegalTypeCheck.checkIdent(IllegalTypeCheck.java:645)
	at com.puppycrawl.tools.checkstyle.checks.coding.IllegalTypeCheck.checkType(IllegalTypeCheck.java:674)
	at com.puppycrawl.tools.checkstyle.utils.TokenUtil.forEachChild(TokenUtil.java:244)
	at com.puppycrawl.tools.checkstyle.checks.coding.IllegalTypeCheck.checkTypeArguments(IllegalTypeCheck.java:719)
	at com.puppycrawl.tools.checkstyle.checks.coding.IllegalTypeCheck.checkType(IllegalTypeCheck.java:675)
	at com.puppycrawl.tools.checkstyle.checks.coding.IllegalTypeCheck.checkClassName(IllegalTypeCheck.java:633)
	at com.puppycrawl.tools.checkstyle.checks.coding.IllegalTypeCheck.visitParameterDef(IllegalTypeCheck.java:563)
	at com.puppycrawl.tools.checkstyle.checks.coding.IllegalTypeCheck.visitToken(IllegalTypeCheck.java:478)
	at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:336)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:407)
	at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:274)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:155)
	at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:98)
	at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:335)
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:294)
	... 5 more
Checkstyle ends with 1 errors.

❯ java $RUN_LOCALE -jar checkstyle-10.3.2-all.jar -c checkstyle-configuration.xml ExampleClass.java
Starting audit...
Audit done.
```

---

Note that any little change to the config file prevents this problem. Even getting rid of the trailing comma or space at the end of StringBuffer in `<property name="illegalClassNames" value="java.lang.StringBuffer, " />` prevents the problem.

---

### PR DESCRIPTION
**Pull request by nrmancuso**
closes #12443

### CODE REVIEW

**Review by romani — APPROVED** [review]
Ok to merge

**Comment by romani** [comment]
Should be merged before Sunday, to release this fix sooner

**Review by rnveach — CHANGES_REQUESTED** [review]
https://github.com/checkstyle/checkstyle/issues/12443#issuecomment-1324375539
I am seeing 3 trims removed, but only 2 fixes here.

**Comment by nrmancuso** [comment]
Third trim does not need to be replaced, trim is called on argument in method.

https://github.com/checkstyle/checkstyle/blob/276263b6493b2b47709407f2315524586619c0a1/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AccessModifierOption.java#L65-L67

**Comment by rnveach** [comment]
> Third trim does not need to be replaced, trim is called on argument in method.

Do we have a similar test to show it is not needed?

**Comment by nrmancuso** [comment]
This is where the other trim was located: https://github.com/checkstyle/checkstyle/commit/308001394030b0ba52921dd8f1b0bef22c90ea83#r90853229

Which calls https://github.com/checkstyle/checkstyle/blob/276263b6493b2b47709407f2315524586619c0a1/src/main/java/com/puppycrawl/tools/checkstyle/checks/naming/AccessModifierOption.java#L65-L67

I can add another tab in config like this: `public\t,\t`

**Comment by nrmancuso** [comment]
@rnveach test case is added

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeWhitespaceInConfig.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+/*
+IllegalType
+validateAbstractClassNames = (default)false
+illegalClassNames = java.lang.StringBuffer,\t
+legalAbstractClassNames = (default)
+ignoredMethodNames = foo1
```
Comment (Vyom-Yadav) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeWhitespaceInConfig.java:6]: Lets keep this is default too:
```
ignoredMethodNames = (default)getEnvironment, getInitialContext
```

**Review by Vyom-Yadav — APPROVED** [review]
One minor, rest LGTM :)

**Comment by rnveach** [comment]
> I can add another tab in config like this: `public\t,\t`, this will cause an unnecessary suppression since this is the test case at https://github.com/checkstyle/checkstyle/pull/12340

There is no test case like this in the other PR.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeWhitespaceInConfig.java (hunk 1 above)
Comment (nrmancuso) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeWhitespaceInConfig.java]: Done

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeWhitespaceInConfig.java]: confirmed.

**Comment by nrmancuso** [comment]
Linking discussion here: https://github.com/checkstyle/checkstyle/pull/12340#discussion_r1031557625

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeWhitespaceInConfig.java (hunk 1 above)
Comment (Vyom-Yadav) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeWhitespaceInConfig.java]: Resolved

**Pull request merged**